plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php

```diff
@@ -156,56 +156,102 @@ private function get_location_context() {
+	}
 
-			$p = new \WP_HTML_Tag_Processor( $block_content );
+	/**
+	 * Attach the init directive to Product Collection block to call
+	 * the onRender callback.
+	 *
+	 * @param string $block_content The HTML content of the block.
+	 * @param string $collection Collection type.
+	 *
+	 * @return string Updated HTML content.
+	 */
+	private function add_rendering_callback( $block_content, $collection ) {
+		$p = new \WP_HTML_Tag_Processor( $block_content );
 
-			// Add `data-wc-navigation-id to the product collection block.
-			if ( $p->next_tag( array( 'class_name' => 'wp-block-woocommerce-product-collection' ) ) ) {
-				$p->set_attribute(
-					'data-wc-navigation-id',
-					'wc-product-collection-' . $this->parsed_block['attrs']['queryId']
-				);
-				$p->set_attribute( 'data-wc-interactive', wp_json_encode( array( 'namespace' => 'woocommerce/product-collection' ), JSON_HEX_TAG | JSON_HEX_APOS | JSON_HEX_QUOT | JSON_HEX_AMP ) );
+		// Add `data-init to the product collection block so we trigger JS event on render.
+		if ( $this->is_next_tag_product_collection( $p ) ) {
+			$p->set_attribute(
+				'data-wc-init',
+				'callbacks.onRender'
+			);
+			if ( $collection ) {
 				$p->set_attribute(
 					'data-wc-context',
 					wp_json_encode(
 						array(
-							// The message to be announced by the screen reader when the page is loading or loaded.
-							'accessibilityLoadingMessage'  => __( 'Loading page, please wait.', 'woocommerce' ),
-							'accessibilityLoadedMessage'   => __( 'Page Loaded.', 'woocommerce' ),
-							// We don't prefetch the links if user haven't clicked on pagination links yet.
-							// This way we avoid prefetching when the page loads.
-							'isPrefetchNextOrPreviousLink' => false,
+							'collection' => $collection,
 						),
 						JSON_HEX_TAG | JSON_HEX_APOS | JSON_HEX_QUOT | JSON_HEX_AMP
 					)
 				);
-				$block_content = $p->get_updated_html();
 			}
+		}
 
-			/**
-			 * Add two div's:
-			 * 1. Pagination animation for visual users.
-			 * 2. Accessibility div for screen readers, to announce page load states.
-			 */
-			$last_tag_position                = strripos( $block_content, '</div>' );
-			$accessibility_and_animation_html = '
+		return $p->get_updated_html();
+	}
+
+	/**
+	 * Attach all the Interactivity API directives responsible
+	 * for client-side navigation.
+	 *
+	 * @param string $block_content The HTML content of the block.
+	 *
+	 * @return string Updated HTML content.
+	 */
+	private function enable_client_side_navigation( $block_content ) {
+		$p = new \WP_HTML_Tag_Processor( $block_content );
+
+		// Add `data-wc-navigation-id to the product collection block.
+		if ( $this->is_next_tag_product_collection( $p ) ) {
+			$p->set_attribute(
+				'data-wc-navigation-id',
+				'wc-product-collection-' . $this->parsed_block['attrs']['queryId']
+			);
+			$current_context = json_decode( $p->get_attribute( 'data-wc-context' ), true ) ?? [];
```

Comment:
This causes a warning when `data-wc-context` is not found. We need to first check before use. Maybe something like this?
`$current_context = ! empty( $p->get_attribute( 'data-wc-context' ) ) ? json_decode( $p->get_attribute( 'data-wc-context' ), true ) : [];`

Reply:
Thanks @roykho for catching it! @wavvves already let me know of this and it's been fixed in https://github.com/woocommerce/woocommerce/pull/50661, pull the newest `trunk` and a warning will be gone!